src/core/util/hugepage_mgr.cpp

```diff
@@ -0,0 +1,258 @@
+#include <cmath>
+#include <cstring>
+#include <mutex>
+#include <fstream>
+#include <sstream>
+
+#include "vlogger/vlogger.h"
+#include "util/sys_vars.h"
+
+#define MODULE_NAME "hugepage_mgr"
+
+hugepage_mgr g_hugepage_mgr;
+
+hugepage_mgr::hugepage_mgr()
+{
+    memset(&m_stats, 0, sizeof(m_stats));
+    m_default_hugepage = read_meminfo("Hugepagesize:");
+    update();
+}
+
+void hugepage_mgr::update()
+{
+    std::lock_guard<decltype(m_lock)> lock(m_lock);
+
+    m_hugepages.clear();
+    read_sysfs();
+}
+
+bool hugepage_mgr::is_hugepage_optimal(size_t hugepage, size_t size)
+{
+    return (size / hugepage) || hugepage_unused_space(hugepage, size) <= HUGEPAGE_UNUSED_OPTIMAL ||
+        hugepage_metric(hugepage, size) <= HUGEPAGE_METRIC_OPTIMAL;
+}
+
+bool hugepage_mgr::is_hugepage_acceptable(size_t hugepage, size_t size)
+{
+    return hugepage_unused_space(hugepage, size) <= HUGEPAGE_UNUSED_ACCEPTABLE ||
+        hugepage_metric(hugepage, size) <= HUGEPAGE_METRIC_ACCEPTABLE;
+}
+
+size_t hugepage_mgr::find_optimal_hugepage(size_t size)
+{
+    std::vector<size_t> hugepages;
+    size_t best_hugepage = 0;
+
+    if (safe_mce_sys().hugepage_log2 == 0) {
+        get_supported_hugepages(hugepages);
+        std::sort(hugepages.begin(), hugepages.end(), std::greater<size_t>());
+    } else {
+        // User requested specific hugepage size - don't check other types.
+        hugepages.push_back(1LU << safe_mce_sys().hugepage_log2);
+    }
+
+    // This is naive algorithm and may work inefficiently in complex scenarios.
+    for (size_t hugepage : hugepages) {
+        if (get_free_hugepages(hugepage) * hugepage >= size) {
+            if (is_hugepage_optimal(hugepage, size)) {
+                return hugepage;
+            }
+            if (is_hugepage_acceptable(hugepage, size) && best_hugepage == 0) {
+                best_hugepage = hugepage;
+            }
+        }
+    }
+
+    if (best_hugepage &&
+        hugepage_unused_space(best_hugepage, size) > HUGEPAGE_UNUSED_WARNING_THRESHOLD) {
+        __log_info_dbg("Allocating %zu bytes with hugepages %zu kB.", size, best_hugepage / 1024U);
+        __log_info_dbg("%zu bytes of the last hugepage may be unused.",
+                       best_hugepage - size % best_hugepage);
+    }
+    return best_hugepage;
+}
+
+void *hugepage_mgr::alloc_hugepages(size_t &size)
+{
+    std::lock_guard<decltype(m_lock)> lock(m_lock);
+
+    size_t hugepage = find_optimal_hugepage(size);
+    size_t hugepage_mask = hugepage - 1;
+    size_t actual_size = size;
+    void *ptr = nullptr;
+
+    if (hugepage) {
+        int map_extra = 0;
+
+        if (hugepage != m_default_hugepage) {
+            map_extra = (int)log2(hugepage) << MAP_HUGE_SHIFT;
+        }
+        actual_size = (size + hugepage_mask) & ~hugepage_mask;
+
+        __log_info_dbg("Allocating %zu bytes with hugepages %zu kB", actual_size, hugepage / 1024U);
+
+        ptr = mmap(NULL, actual_size, PROT_READ | PROT_WRITE,
```

Comment:
is mmap always allocates complete hugepages or it can allocate memory from a page that is not completely used?

Reply:
mmap(MAP_HUGETLB) always allocates complete pages. So address is guaranteed to be aligned to the hugepage boundary and requested length is aligned if needed. However, we align length explicitly to (1) be able utilize the tail and (2) munmap() requires aligned length argument.